### PR TITLE
Add label smoothing for cross entropy

### PR DESCRIPTION
**Pull request by Tcc0403**
## Summary
Aim to solve #81.

### For loss:
Label smoothing regularization ( LSR ) by replacing the label distribution $q(k) = \delta_{k,y}$ with
```math
q'(k) = (1 - \epsilon)\delta_{k,y} + \frac{\epsilon}{K}
```
Considering cross entropy with LSR is

```math
\begin{align}
L' = H(q', p) &= -\sum^K_{k=1}log\ {p(k)}q'(k) = -\sum^K_{k=1}log\ {p(k)}((1 - \epsilon)\delta_{k,y} + \frac{\epsilon}{K})\\
                    &=  -\sum^K_{k=1}log\ {p(k)}(1 - \epsilon)q(k)   -\sum^K_{k=1}log\ {p(k)}\frac{\epsilon}{K} \\
                    &= (1 - \epsilon)H(q,p) + \frac{\epsilon}{K} \sum^K_{k=1} log\ softmax(x_k)\\
                    &= (1- \epsilon)L + \frac{\epsilon}{K}\ SmoothLoss,

\end{align}
```
where $L = H(q,p)$ is the original loss and $\sum^K_{k=1} log\ softmax(x_k)$ is smooth loss.

### For gradients:
The original:
```math
\begin{align}
\frac{\partial L}{\partial x_i} &= p(k) - q(k)\\
                                            &= \begin{cases}
                                                   softmax(x_i) ,                        &  i \neq y \\
                                                   softmax(x_i) - 1,                    &  i = y
                                                   
\end{cases}
\end{align} 
```
With LSR:
```math
\begin{align}
\frac{\partial L'}{\partial x_i} &= p(k) - q'(k)\\
                                            &= softmax(x_i) - (1 - \epsilon)\delta_{k,y} + \frac{\epsilon}{K}\\
                                            &= \begin{cases} softmax(x_i) - \frac{\epsilon}{K},                        &  i \neq y \\
                                                   softmax(x_i) - \frac{\epsilon}{K} - (1 - \epsilon) &  i = y

\end{cases}
\end{align}
```

We can handle the $i = y$ case by simply adding $-(1-\epsilon)$ after computing all $i$.

Reference:
[Rethinking the Inception Architecture for Computer Vision](https://arxiv.org/abs/1512.00567)

## Testing Done
Add a unit test for label smoothing.

- Hardware Type: RTX-3080
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence
```bash
❯ python3 -m pytest test/transformers/test_cross_entropy.py
============================================ test session starts =============================================
platform linux -- Python 3.10.12, pytest-8.3.2, pluggy-1.5.0
rootdir: /home/tcc/Liger-Kernel
collected 94 items

test/transformers/test_cross_entropy.py .............................................................. [ 65%]
...............................F                                                                       [100%]

================================================== FAILURES ==================================================
__________________________________ test_large_no_exception[8-16384-128256] ___________________________________

B = 8, T = 16384, V = 128256

    @pytest.mark.parametrize(
        "B, T, V",
        [
            (
                8,
                8192,
                128256,
            ),  # _input = 16GB, total = ~32GB, 8405385216 > 2,147,483,647, so we need int64
            (8, 16384, 128256),  # _input = 32GB, total = ~64GB
        ],
    )
    # @pytest.mark.skipif(
    #     torch.cuda.get_device_properties(0).total_memory < 64 * 1000 * 1000 * 1000,
    #     reason="Needs 64GB+ GPU memory.",
    # )
    def test_large_no_exception(B, T, V):
        # The large inputs were hitting cuda illegal memory access because of
        # https://github.com/triton-lang/triton/issues/1058
>       _full_pass_once(B, T, V)

test/transformers/test_cross_entropy.py:401:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

B = 8, T = 16384, V = 128256

    def _full_pass_once(B, T, V):
        torch.manual_seed(0)
        liger_ce = LigerCrossEntropyLoss()

>       _input = torch.randn(
            B * T, V, requires_grad=True, device="cuda", dtype=torch.bfloat16
        )
E       torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 31.31 GiB. GPU 0 has a total capacity of 10.00 GiB of which 8.84 GiB is free. Including non-PyTorch memory, this process has 17179869184.00 GiB memory in use. Of the allocated memory 0 bytes is allocated by PyTorch, and 0 bytes is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

test/transformers/test_cross_entropy.py:374: OutOfMemoryError
========================================== short test summary info ===========================================
FAILED test/transformers/test_cross_entropy.py::test_large_no_exception[8-16384-128256] - torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 31.31 GiB. GPU 0 has a total capacity of 10...
================================== 1 failed, 93 passed in 130.88s (0:02:10) ==================================
```
```bash
❯ make test
python -m pytest --disable-warnings test/ --ignore=test/convergence
============================================ test session starts =============================================
platform linux -- Python 3.10.12, pytest-8.3.2, pluggy-1.5.0
rootdir: /home/tcc/Liger-Kernel
collected 256 items

test/transformers/test_auto_model.py .                                                                 [  0%]
test/transformers/test_cross_entropy.py ssssssssssssssssssssssss............ssssssssssssssssssssssssss [ 24%]
ssssssssssssssssssssssssssssssss                                                                       [ 37%]
test/transformers/test_embedding.py ...........                                                        [ 41%]
test/transformers/test_fused_linear_cross_entropy.py ................                                  [ 47%]
test/transformers/test_geglu.py ............                                                           [ 52%]
test/transformers/test_layer_norm.py ................                                                  [ 58%]
test/transformers/test_monkey_patch.py .....                                                           [ 60%]
test/transformers/test_rms_norm.py ............................................................        [ 83%]
test/transformers/test_rope.py ..................                                                      [ 91%]
test/transformers/test_swiglu.py ....................                                                  [ 98%]
test/transformers/test_trainer_integration.py .                                                        [ 99%]
test/triton/test_triton_monkey_patch.py ..                                                             [100%]

================================ 174 passed, 82 skipped in 123.06s (0:02:03) =================================
```
```bash
❯ make checkstyle
flake8 .; flake8_status=$?; \
isort .; isort_status=$?; \
black .; black_status=$?; \
if [ $flake8_status -ne 0 ] || [ $isort_status -ne 0 ] || [ $black_status -ne 0 ]; then \
        exit 1; \
fi
Skipped 2 files
All done! ✨ 🍰 ✨
68 files left unchanged.
```
```bash
❯ make test-convergence
HF_DATASETS_OFFLINE=1 python -m pytest --disable-warnings test/convergence
============================================ test session starts =============================================
platform linux -- Python 3.10.12, pytest-8.3.2, pluggy-1.5.0
rootdir: /home/tcc/Liger-Kernel
collected 30 items

test/convergence/test_mini_models.py ..............                                                    [ 46%]
test/convergence/test_mini_models_no_logits.py ................                                        [100%]

======================================= 30 passed in 223.18s (0:03:43) =======================================
```

### CODE REVIEW

**Comment by ByronHsu** [comment]
nice work! cc @shivam15s

**Comment by Tcc0403** [comment]
Current PR can't pass tests with ignore_index and label_smoothing both on. I'm working on it.

File: src/liger_kernel/ops/cross_entropy.py [hunk 1]
```diff
@@ -73,16 +75,36 @@ def liger_cross_entropy_kernel(
         d = d * tl.exp(m - m_new) + tl.sum(tl.exp(X_block - m_new))
         m = m_new
 
+    # we need to compute sum(log(softmax(x_i))) for smooth_loss
+    smooth_loss = 0.0
+    log_d = tl.log(d)  # avoid redundant calculations
```
Comment (Tcc0403) [inline src/liger_kernel/ops/cross_entropy.py:80]: not sure if the compiler can optimize it

File: src/liger_kernel/ops/cross_entropy.py [hunk 2]
```diff
@@ -73,16 +75,36 @@ def liger_cross_entropy_kernel(
...
+
     # 4. [Online softmax] second pass: calculate the gradients
     # dx_y = (softmax(x_y) - 1) / N
     # dx_i = softmax(x_i) / N, i != y
     # N is the number of non ignored elements in the batch
+    # For label smoothing:
+    # dx_i = (softmax(x_y) - label_smoothing / V) / N, V = n_cols, i != y
+    # dx_y = (softmax(x_y) - label_smoothing / V - (1 - label_smoothing)) / N
+    #      = dx_i - (1 - label_smoothing) / N
+    eps = label_smoothing / n_cols
```
Comment (Tcc0403) [inline src/liger_kernel/ops/cross_entropy.py:101]: concerned about naming

Comment (Tcc0403) [reply src/liger_kernel/ops/cross_entropy.py:101]: Comments for label smoothing version are identical to non label smoothing ones.
Not sure which one we should keep, keeping both for now.

**Comment by Tcc0403** [comment]
Ready for review!

File: src/liger_kernel/ops/cross_entropy.py [hunk 3]
```diff
@@ -225,11 +256,14 @@ def forward(ctx, _input, target, ignore_index):
         _input (tensor): The input tensor of shape (BT, V) where B is batch size, T is sequence length, V is vocab size.
         target (tensor): The target tensor of shape (BT) where each value is in [0, V-1].
         ignore_index (int): The index to ignore in the target.
+        label_smoothing (float): A float in [0.0, 1.0].
```
Comment (yundai424) [inline src/liger_kernel/ops/cross_entropy.py:259]: minor: could we also add `"The amount of smoothing when computing the loss, where 0.0 means no smoothing."` here because this will be sort of an open API

**Review by yundai424 — COMMENTED** [review]
love the detailed comments here!

File: src/liger_kernel/ops/cross_entropy.py (hunk 3 above)
Comment (Tcc0403) [reply src/liger_kernel/ops/cross_entropy.py]: added!

**Comment by Tcc0403** [comment]
> Should I also add a unit test of passing both ignore_index and label_smoothing to ensure there's no conflict?


Btw this was the unit test I mentioned. I wasn't sure if this test makes sense. 
Now I think about it, in `ignore_index==target` case, the only thing cross entropy kernel does is setting gradients to 0,  so it's unlikely that both parameters will cause any conflict. It's more like a compound test to ensure both parameters can work.

TL;DR: we don't need this test.
```python
def _test_correctness_with_ignore_index_and_label_smoothing_once(
    target_ce, B, T, V, ignore_index, label_smoothing, scalar, dtype, atol, rtol
):
    torch.manual_seed(0)
    torch_ce = CrossEntropyLoss(
        ignore_index=ignore_index, label_smoothing=label_smoothing
    )

    _tensor = torch.randn(B * T, V, device="cuda", dtype=dtype) * scalar
    _input = _tensor.detach().clone().requires_grad_(True)
    _input2 = _tensor.detach().clone().requires_grad_(True)

    target = torch.randint(0, V, (B * T,), device="cuda", dtype=torch.long)

    # Assign some random number of elements as ignore_index
    num_elements_to_assign = torch.randint(
        1, B * T // 2, (1,)
    ).item()  # Random number of elements to set to ignore_index
    indices_to_assign = torch.randperm(B * T)[
        :num_elements_to_assign
    ]  # Randomly select indices
    target[indices_to_assign] = ignore_index

    output = torch_ce(_input, target)
    output2 = target_ce(_input2, target)

    assert torch.allclose(output, output2, atol=atol, rtol=rtol)

    output.backward()
    output2.backward()
    assert torch.allclose(_input.grad, _input2.grad, atol=atol, rtol=rtol)
```

**Comment by lancerts** [comment]
> Should I also add a unit test of passing both ignore_index and label_smoothing to ensure there's no conflict?

Yes, please add, ty

**Comment by Tcc0403** [comment]
> > Should I also add a unit test of passing both ignore_index and label_smoothing to ensure there's no conflict?
> 
> Yes, please add, ty

Will do.

**Review by ByronHsu — COMMENTED** [review]
We also need to expose label_smoothing [here](https://github.com/linkedin/Liger-Kernel/blob/5151712ea17fc9aa595533cb2d913b42b774c8b3/src/liger_kernel/ops/fused_linear_cross_entropy.py#L174)

File: test/transformers/test_cross_entropy.py [hunk 4]
```diff
@@ -61,6 +61,61 @@ def _test_correctness_with_ignore_index_once(
...
+    output2 = target_ce(_input2, target)
+
+    assert torch.allclose(output, output2, atol=atol, rtol=rtol)
+
+    output.backward()
+    output2.backward()
+    assert torch.allclose(_input.grad, _input2.grad, atol=atol, rtol=rtol)
+
+
+def _test_correctness_with_parameters_once(
```
Comment (ByronHsu) [inline test/transformers/test_cross_entropy.py:86]:
```suggestion
def _test_correctness_with_label_smoothing_with_ignore_index_once(
```

**Review by ByronHsu — COMMENTED** [review]
super solid work! one of the best prs we have received so far :)

**Comment by ByronHsu** [comment]
Also, i think because @qingquansong has refactored out the fn body from autograd fn. [Here](https://github.com/linkedin/Liger-Kernel/blob/5151712ea17fc9aa595533cb2d913b42b774c8b3/src/liger_kernel/ops/fused_linear_cross_entropy.py#L73) we can just call cross_entropy_forward, which is cleaner than calling kernel directly (can be another PR)

File: src/liger_kernel/ops/cross_entropy.py [hunk 5]
```diff
@@ -63,12 +65,18 @@ def liger_cross_entropy_kernel(
         X_ptr + y
     )  # we need to store the original value of X_y for the loss calculation
 
+    scaled_x_sum = 0.0
```
Comment (ByronHsu) [inline src/liger_kernel/ops/cross_entropy.py:68]:
```suggestion
    # Label smoothing is a general case of normal cross entropy
    # See the full derivation at https://github.com/linkedin/Liger-Kernel/pull/198#issue-2503665310
    scaled_x_sum = 0.0
```

File: src/liger_kernel/ops/cross_entropy.py [hunk 6]
```diff
@@ -97,9 +109,19 @@ def liger_cross_entropy_kernel(
     # So we can safely calculate log (softmax(X_y)) without overflow
     loss = -(ori_X_y - m - tl.log(d))
 
-    # 6. Specially handle the i==y case where `dx_y = (softmax(x_y) - 1) / N`
+    # Orginal loss = H(q, p),  with label smoothing regularization = H(q', p) and (label_smoothing / V) = eps
+    # H(q', p) = (1 - label_smoothing) * H(q, p) + label_smoothing * H(u, p)
+    #          = (1 - label_smoothing) * H(q, p) + eps * sum(softmax(x_i))
```
Comment (ByronHsu) [inline src/liger_kernel/ops/cross_entropy.py:114]:
```suggestion
    #          = (1 - label_smoothing) * H(q, p) + eps * sum(logsoftmax(x_i))
    # By using m (global max of xi) and d (sum of e^(xi-m)), we can simplify as:
```

File: src/liger_kernel/ops/cross_entropy.py [hunk 7]
```diff
@@ -97,9 +109,19 @@ def liger_cross_entropy_kernel(
...
+    #          = (1 - label_smoothing) * H(q, p) + (-sum(x_i * eps) + label_smoothing * (m + logd))
+    # Refer to H(q', p) in section 7 of the paper: https://arxiv.org/pdf/1512.00567
+    # pytorch: https://github.com/pytorch/pytorch/blob/2981534f54d49fa3a9755c9b0855e7929c2527f0/aten/src/ATen/native/LossNLL.cpp#L516
```
Comment (ByronHsu) [inline src/liger_kernel/ops/cross_entropy.py:117]:
```suggestion
    # pytorch: https://github.com/pytorch/pytorch/blob/2981534f54d49fa3a9755c9b0855e7929c2527f0/aten/src/ATen/native/LossNLL.cpp#L516
    # See full derivation at https://github.com/linkedin/Liger-Kernel/pull/198#issuecomment-2333753087
```

**Review by ByronHsu — COMMENTED** [review]
super solid work! one of the best prs we have received so far :)

**Review by ByronHsu — APPROVED** [review]
